pyrealm/demography/canopy_functions.py

```diff
@@ -0,0 +1,415 @@
+) -> NDArray[np.float32]:
+    r"""Calculate height of maximum crown radius.
+
+    The height of the maximum crown radius (:math:`z_m`) is derived from the canopy
+    shape parameters (:math:`m,n`) and the resulting fixed proportion (:math:`p_{zm}`)
+    for plant functional types. These shape parameters are defined as part of the
+    extension of the T Model presented by :cite:t:`joshi:2022a`.
+
+    The value :math:`z_m` is the height above ground where the largest canopy radius is
+    found, given the proportion and the estimated stem height (:math:`H`) of
+    individuals.
+
+    .. math::
+
+        z_m = p_{zm} H
+
+    Args:
+        z_max_prop: Canopy shape parameter of the PFT
+        stem_height: Stem height of individuals
+    """
+    """Calculate z_m, the height of maximum crown radius."""
+
+    return stem_height * z_max_prop
+
+
+def calculate_canopy_r0(
+    q_m: NDArray[np.float32], crown_area: NDArray[np.float32]
+) -> NDArray[np.float32]:
+    r"""Calculate scaling factor for height of maximum crown radius.
+
+    This scaling factor (:math:`r_0`) is derived from the canopy shape parameters
+    (:math:`m,n,q_m`) for plant functional types and the estimated crown area
+    (:math:`A_c`) of individuals. The shape parameters are defined as part of the
+    extension of the T Model presented by :cite:t:`joshi:2022a` and :math:`r_0` is used
+    to scale the crown area such that the crown area at the  maximum crown radius fits
+    the expectations of the T Model.
+
+    .. math::
+
+        r_0 = 1/q_m  \sqrt{A_c / \pi}
+
+    Args:
+        q_m: Canopy shape parameter of the PFT
+        crown_area: Crown area of individuals
+    """
+    # Scaling factor to give expected A_c (crown area) at
+    # z_m (height of maximum crown radius)
+
+    return 1 / q_m * np.sqrt(crown_area / np.pi)
+
+
+def _validate_z_args(z: NDArray[np.float32], *args: NDArray[np.float32]) -> None:
+    """Shared validation routine for vertical height arguments.
+
+    The canopy functions that take a height argument ``z`` can take a range of input
+    shapes for ``z``, alongside a set of other row arrays representing cohort
+    properties. This function is used to check that the cohort properties are of equal
+    length and that the ``z`` value has one of the accepted shapes described in
+    :meth:`~pyrealm.demography.canopy_functions.calculate_relative_canopy_radius_at_z`.
+
+    Args:
+        z: The inputs to the ``z`` argument of a function.
+        args: Other arrays representing cohort properties.
+    """
+
+    if z.size == 1 or z.ndim == 1:
+        # All same height or stem specific heights - check z is either a scalar or also
+        # a row vector of the right length.
+        check_input_shapes(z, *args)
+        return
+    elif z.ndim == 2 and z.shape[1] == 1:
+        # Z is a column vector, just check stem properties.
+        check_input_shapes(*args)
+        return
+
+    raise ValueError("Invalid shape for the z value.")
```

Comment:
Yeah - that's more of an insult than an Exception message.

I think those validation functions probably need to be unpacked a bit - the shape requirements are intricate and we'd expect most users to interact with higher level APIs but the functions should be more helpful in their failure modes than "haha luzer".

Reply:
OK - that kicked off a major rewrite of the validation functions and tests. There is now a single integrated validation test rather than two separate ones, that does a much clearer job of setting out the valid combinations of shapes and providing specific guidance in failure cases. The test suite got a lot more clear too, I think.